Add Starkli-Scarb interaction using katana (#216)
dojoengine/book: docs/pages/toolchain/katana/interact.md

# Interact with Katana: Deploy a simple Cairo smart contract using Katana

Welcome to this tutorial where we'll guide you through deploying a raw cairo smart using katana as a local devnet. In order to accomplish this we have to install the following tools.

## Install Starkli, Scarb and Katana

To install `Starkli`, open a new terminal

```bash
curl https://get.starkli.sh | sh
starkliup
```

You can check your installation by running `starkli --version`, then you will get the starkli version.

To install `Scarb`

```bash
curl --proto '=https' --tlsv1.2 -sSf https://docs.swmansion.com/scarb/install.sh | sh
```

You can check your installation by running `scarb --version`, then you will get the starkli version.

To install `Katana`, use the `dojoup` installer from the command line:

> ```bash

Reviewer: could you move these to a 'snippets' directory and we can reuse the snippet everywhere. https://vocs.dev/docs/guides/code-snippets

Also could you do it for other CLI commands that are generalised

curl -L https://install.dojoengine.org | bash
dojoup -v 0.6.0-alpha.5
```

You can check your installation by running `katana --version`, then you will get the katana version.

## Basics of Katana and Starkli

### katana sequencer

```console
katana --disable-fee
```

> After starting the node, a list of accounts will be automatically generated and deployed.

Reviewer: ```suggestion
After starting the Katana, a list of accounts will be automatically generated and deployed.
```

### Starkli built-in accounts and configuration

Starkli supports a list of built-in accounts for katana. These built-in accounts are for local development and no one should be using them for anything serious anyways. For example the address of the `Katana-0` built-in account is `0x6162896d1d7ab204c7ccac6dd5f8e9e7c25ecd5ae4fcb4ad32e57786bb46e03`. You can check the full list of account addresses [here](https://github.com/xJonathanLEI/starkli/blob/e9a28f1b6e37bcc9fc53b7b7130e935894856739/src/account.rs#L76).

We will need this `katana-0` built-in account later when we define our `STARKNET_ACCOUNT`. For more details about accounts, please refer to the accounts [page](https://book.starkli.rs/accounts).

> ## Contract Deployment and Interaction

Reviewer: add context here that this is a new directory.

### Create a Simple Storage project

```bash
scarb new simple_storage
```

Add contract dependencies to scarb.toml

```toml
[dependencies]
starknet = "2.5.4"

[[target.starknet-contract]]
```

Copy the simple storage contract to lib.cairo

```rust,ignore
#[starknet::interface]
trait ISimpleStorage<TContractState> {
    fn set(ref self: TContractState, x: u128);
    fn get(self: @TContractState) -> u128;
}

#[starknet::contract]
mod SimpleStorage {
    use starknet::get_caller_address;
    use starknet::ContractAddress;

    #[storage]
    struct Storage {
        stored_data: u128
    }

    #[abi(embed_v0)]
    impl SimpleStorage of super::ISimpleStorage<ContractState> {
        fn set(ref self: ContractState, x: u128) {
            self.stored_data.write(x);
        }
        fn get(self: @ContractState) -> u128 {
            self.stored_data.read()
        }
    }
}
```

> ### Compile contract and Environment variables setup

Reviewer: ```suggestion
### Compile contract and environment variables setup
```

> ### Compile contract and Environment variables setup

Reviewer: ```suggestion
### Compile contract and add environment variables
```

Compile your contract using scarb

```bash
scarb build
```

> Having compiled the smart contract, it's time to declare it with Starkli and katana. For convenient management, place the following environment variables in a .env file within the `src/` directory.

Reviewer: ```suggestion
Having compiled the smart contract, it's time to declare it with Starkli and katana. For clean enviroment management, place the following environment variables in a .env file within the `src/` directory.
```

```bash
export STARKNET_ACCOUNT=katana-0        #A pre-funded account on the local development network.
export STARKNET_RPC=http://0.0.0.0:5050 #To specify the network, targeting the local katana devnet.
```

Then, ensure your project acknowledges the environment variables:

```bash
source .env
```

These settings significantly streamline Starkli command operations, ensuring a smoother and more efficient workflow.

### Declare contract

Make sure Katana is already running in separate terminal. Otherwise launch katana

```bash
katana --disable-fee
```

To declare your contract, execute:

```bash
starkli declare target/dev/simple_storage_SimpleStorage.contract_class.json
```

Upon successful command execution, you'll obtain a contract class hash: This unique hash serves as the identifier for your contract class within Starknet. For example:

```sh

> Class hash declared: 0x07ad2516dd66fb2e274e78d4357837cad689c9fffaa347feb9800b231b37b306

Reviewer: do the full printout here and focus on the class hash

```

### Deploy contract

```bash
starkli deploy <class_hash_of_the_contract_to_be_deployed>
```

For this contract we did not specified a `constructor` function, thus we don't need to pass any constructor argument

````bash
starkli deploy 0x07ad2516dd66fb2e274e78d4357837cad689c9fffaa347feb9800b231b37b306

After running, expect an output similar to:

```console
Deploying class 0x07ad2516dd66fb2e274e78d4357837cad689c9fffaa347feb9800b231b37b306 with salt 0x02c93ad00ce6f894729baeafd1fd0456c5a5c540c1caa053ab5392f27ea8f130...
The contract will be deployed at address 0x03da69257a94a06a1101c1413d78551e38d91ca180c0fc26004650a427238f4e
Contract deployment transaction: 0x06817bc837ce4df879fe903ca4700a860ce8165742bd74bdadf379618e89cccd
Contract deployed:
0x03da69257a94a06a1101c1413d78551e38d91ca180c0fc26004650a427238f4e
````

### Call contract [only read state]

The first parameter is the contract address, the second parameter is the function to be called.

```bash
starkli call 0x03da69257a94a06a1101c1413d78551e38d91ca180c0fc26004650a427238f4e get
```

After running, expect an output similar to:

```console
[
    "0x0000000000000000000000000000000000000000000000000000000000000000"
]
```

> It means the vale of `stored_data` is cero.

Reviewer: ```suggestion
It means the vale of `stored_data` is zero.
```

To facilitate the seamless execution of Starkli commands, it is crucial to configure specific environment variables. Two primary variables are essential for this purpose:

##### Account Configuration:

Create an environment variable for your account, specifically for a pre-funded account on the local development network.

```bash
export STARKNET_ACCOUNT=katana-0
```

##### Network Designation:

Establish another environment variable to specify the network, particularly targeting the local katana devnet.
bash

```bash
export STARKNET_RPC=http://0.0.0.0:5050
```

### Contract Deployment and Interaction

1. Create a Vote project

```console
scarb new vote
```

Add contract dependencies to scarb.toml

```toml
[dependencies]
starknet = "2.5.0"

> [[target.starknet-contract]]

Reviewer 2: ```suggestion

[[target.starknet-contract]]
```

```

> Copy the vote contract to lib.cairo

Reviewer 2: This contract is interesting, but is huge IMHO.
What do you think about having a simple counter or something like this? Or perhaps you've some other cases you want to add?

Author: You are right. I suggest we can show a simple/basic example with a `counter contract`. Then we also can show an intermediate example with this `voting contract`

Reviewer 2: Yep sounds good! The idea is as some reader may just dive into starknet, having something simpler first would not discourage them. :)

            assert(can_vote == true, 'USER_ALREADY_VOTED');
        }
    }

    /// @dev Implement the VotingResultTrait for the Vote contract
    #[generate_trait]
    impl VoteResultFunctionsImpl of VoteResultFunctionsTrait {
        // @dev Internal function to get the voting results (yes and no vote counts)
        fn _get_voting_result(self: @ContractState) -> (u8, u8) {
            let n_yes: u8 = self.yes_votes.read();
            let n_no: u8 = self.no_votes.read();

            (n_yes, n_no)
        }

        // @dev Internal function to calculate the voting results in percentage
        fn _get_voting_result_in_percentage(self: @ContractState) -> (u8, u8) {
            let n_yes: u8 = self.yes_votes.read();
            let n_no: u8 = self.no_votes.read();

            let total_votes: u8 = n_yes + n_no;

            if (total_votes == 0_u8) {
                return (0, 0);
            }
            let yes_percentage: u8 = (n_yes * 100_u8) / (total_votes);
            let no_percentage: u8 = (n_no * 100_u8) / (total_votes);

            (yes_percentage, no_percentage)
        }
    }
}
```

2. Compile contract

```bash
scarb build
```

3. Environment variables

>    Having compiled the smart contract, it's time to declare it with Starkli and katana. For convenient management, place the following environment variables in a .env file within the `src/` directory.

Reviewer 2: We've already defined those variables at the beginning of the tutorial. Seems a better approach to actually use the `.env`. We may refer to the starkli book for the first part then, and keep this one as it's more intuitive and re-usable?

Author: Yes. I was thinking how to explain well and where to keep the environment variables. I am gonna follow your suggestion on this.

```bash
starkli invoke 0x02c44f2d396fc5f9caa551e8c1d901d943a3b8cc5c433c88a1bf10b1f15fcd15 vote 1
```

Now let's vote `No` with `katana-1` user

```bash
starkli invoke 0x02c44f2d396fc5f9caa551e8c1d901d943a3b8cc5c433c88a1bf10b1f15fcd15 vote 0 --account katana-1
```

Let's try to vote again with `katana-0` user

```bash
starkli invoke 0x02c44f2d396fc5f9caa551e8c1d901d943a3b8cc5c433c88a1bf10b1f15fcd15 vote 0
```

Since the same user/signer cannot vote repeatedly Katana will report an error.

```console
Transaction execution error: "Error in the called contract (0x06162896d1d7ab204c7ccac6dd5f8e9e7c25ecd5ae4fcb4ad32e57786bb46e03):
Error at pc=0:4573:
Got an exception while executing a hint: Hint Error: Execution failed. Failure reason: 0x555345525f414c52454144595f564f544544 ('USER_ALREADY_VOTED').
Cairo traceback (most recent call last):
Unknown location (pc=0:67)
Unknown location (pc=0:1835)
Unknown location (pc=0:2478)
Unknown location (pc=0:3255)
Unknown location (pc=0:3795)

Error in the called contract (0x02c44f2d396fc5f9caa551e8c1d901d943a3b8cc5c433c88a1bf10b1f15fcd15):
Execution failed. Failure reason: 0x555345525f414c52454144595f564f544544 ('USER_ALREADY_VOTED').

> ```

Reviewer 2: That's nice to show a failure. :+1:
